tracer/src/Datadog.Trace/Iast/InsecureAuthAnalyzer.cs

```diff
@@ -0,0 +1,115 @@
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(InsecureAuthAnalyzer));
+
+#if !NETFRAMEWORK
+    public static void AnalyzeInsecureAuth(IHeaderDictionary headers, IntegrationId integrationId, int statusCode)
+    {
+        try
+        {
+            // Check for Authorization header
+            const string headerName = HeaderNames.Authorization;
+            if (!headers.TryGetValue(headerName, out var authHeader))
+            {
+                return;
+            }
+
+            if (IgnoredStatusCode(statusCode))
+            {
+                return;
+            }
+
+            var detectedScheme = InsecureSchemeDetected(authHeader);
+            if (detectedScheme is null)
+            {
+                return;
+            }
+
+            IastModule.OnInsecureAuthProtocol($"{headerName}: {detectedScheme}", integrationId);
+        }
+        catch (Exception error)
+        {
+            Log.Error(error, $"{nameof(InsecureAuthAnalyzer)}.{nameof(AnalyzeInsecureAuth)}.netcore exception");
+        }
+    }
+#endif
+
+#if NETFRAMEWORK
+    public static void AnalyzeInsecureAuth(System.Collections.Specialized.NameValueCollection headers, IntegrationId integrationId, int statusCode)
+    {
+        try
+        {
+            // Check for Authorization header
+            const string headerName = "Authorization";
+            var authHeader = headers[headerName];
+            if (string.IsNullOrEmpty(authHeader))
+            {
+                return;
+            }
+
+            if (IgnoredStatusCode(statusCode))
+            {
+                return;
+            }
+
+            var detectedScheme = InsecureSchemeDetected(authHeader);
+            if (detectedScheme is null)
+            {
+                return;
+            }
+
+            IastModule.OnInsecureAuthProtocol($"{headerName}: {detectedScheme}", integrationId);
+        }
+        catch (Exception error)
+        {
+            Log.Error(error, $"{nameof(InsecureAuthAnalyzer)}.{nameof(AnalyzeInsecureAuth)}.net461 exception");
+        }
+    }
+#endif
+
+    private static bool IgnoredStatusCode(int statusCode)
+    {
+        // To minimize false positives when we get auth credentials to a page that doesn't exist
+        // (e.g. happens with vulnerability scanners), we'll just ignore this vulnerability when
+        // there is no success response.
+        return statusCode >= 400;
+    }
+
+    private static string? InsecureSchemeDetected(string authHeader)
+    {
+        var insecureSchemes = new[] { "Basic", "Digest" };
+
+        // The auth header can be a concatenation of multiple Authorization headers (concatenated by a comma)
+        // This is not a standard practice for the Authorization header but can theoretically happen.
+        var elements = authHeader.Split(',');
+        foreach (var element in elements)
+        {
+            var scheme = element.Trim(' ').Split(' ')[0];
```

Comment:
Not a fan of all the extra garbage here 🙁 This is running for every single request, so we should probably try to optimize it if possible, though I'm not sure what's best tbh

Reply:
changed in b1e18bd

I changed it to a trim at the start and doing a startsWith comparison. I don't know how it can be better optimised.

Reply:
Well after more research, I came to this benchmark:

```csharp
namespace Benchmarks.Trace.microbenchmarks
{
    public class InsecureAuthBenchmark
    {
        private const string AuthHeader = "      baSic dG9tOm5ldG1hc2s=";
        private static Regex BasicRegex = new Regex("^(?i)Basic", RegexOptions.Compiled | RegexOptions.IgnoreCase);
            
        [Benchmark]
        public void TrimInsecureAuthBasicDetection()
        {
            var trimmedElement = AuthHeader.TrimStart();
            if (trimmedElement.StartsWith("Basic", StringComparison.OrdinalIgnoreCase))
            {
                return;
            }
        }
        
        [Benchmark]
        public void RegexInsecureAuthBasicDetection()
        {
            if (BasicRegex.IsMatch(AuthHeader))
            {
                return;
            }
        }
        
        [Benchmark]
        public void CustomDetectionNoAllocationAuthBasicDetection()
        {
            // 1 - Trim whitespaces from the beginning of the string
            var i = 0;
            while (i < AuthHeader.Length && char.IsWhiteSpace(AuthHeader[i]))
            {
                i++;
            }
            
            // 2 - Check if the string starts with "Basic" (case insensitive)
            if (i + 5 < AuthHeader.Length &&
                (AuthHeader[i] == 'B' || AuthHeader[i] == 'b') &&
                (AuthHeader[i + 1] == 'A' || AuthHeader[i + 1] == 'a') &&
                (AuthHeader[i + 2] == 'S' || AuthHeader[i + 2] == 's') &&
                (AuthHeader[i + 3] == 'I' || AuthHeader[i + 3] == 'i') &&
                (AuthHeader[i + 4] == 'C' || AuthHeader[i + 4] == 'c'))
            {
                return;
            }
        }
    }
}
```

With the result:
```
|                                        Method |      Mean |     Error |    StdDev |
|---------------------------------------------- |----------:|----------:|----------:|
|                TrimInsecureAuthBasicDetection | 12.505 ns | 0.2720 ns | 0.2910 ns |
|               RegexInsecureAuthBasicDetection | 19.156 ns | 0.3698 ns | 0.3460 ns |
| CustomDetectionNoAllocationAuthBasicDetection |  5.176 ns | 0.0375 ns | 0.0313 ns |
```

Reply:
changed in 116ce80